keps/sig-auth/2799-reduction-of-secret-based-service-account-token/README.md

```diff
@@ -0,0 +1,410 @@
+
+### Goals
+
+- No auto-generation of secret-based service account token.
+- Removal of unused auto-generated secret-based service account tokens
+
+### Non-Goals
+
+## Proposal
+
+- In release 1.23, change the service account control loop in Token Controller
+  to not create secret for the service account. At the same time, warn usage of
+  auto-created secret-based service account tokens and enrouage users to use
+  TokenRequest API or manually-created secret-based service account tokens.
+- In release 1.25, purge auto-generated secret-based service account tokens.
+
+### Notes/Constraints/Caveats
+
+- A warning mechanism should be implemented to help users
+  migrate.
+- Only clean up auto-generated tokens which:
+  - are not referenced by pods
+  - have not been used to authenticate for some duration (time duration or number of releases)
+- To consult active usage of secret-based tokens, metric
+  `serviceaccount_legacy_tokens_total` or audit annotation
+  `authentication.k8s.io/legacy-token` could be used.
+
+### Risks and Mitigations
+
+- In release 1.24 (LegacyServiceAccountTokenNoAutoGeneration Beta), consumers
+  depending directly on waiting for and reading tokens out of auto-generated
+  secrets might stop working. To mitigate,
+  1. In 1.23, announce the stopping auto-generating token secrets will happen in
+     1.24
+  2. Publish pointers to TokenRequest or the manual secret request flow.
+- In release 1.25 (LegacyServiceAccountTokenCleanUp Beta), usage of secret-based
+  token that are referenced in service accounts might stop working. To mitigate,
+  add poiners of TokenRequest API and manually created tokens in the validation
+  result.
+
+## Design Details
+
+### Token Controller
+
+#### LegacyServiceAccountTokenNoAutoGeneration:
+
+1. Token Controller stops to create secret for service accounts.
+2. To falicitate LegacyServiceAccountTokenCleanUp, when this feature
+   is Beta in 1.24 where pod-based usage of auto-generated tokens that are injected
+   by admission controller should have dried up,
+   - Annotate auto-generated secrets with a date indicating when we started
+     tracking use (e.g."secret.kubernetes.io/used-since: \$date", default to "never").
+   - When a legacy token is used, if it is from a secret with the `used-since`
+     annotation, issue a warning, update the value, and record in a metric.
+     Optionally, add a label `in-use` for fast query.
+
+#### LegacyServiceAccountTokenCleanUp
+
+Token Controller starts to remove unused auto-generated secrets (secrets
+bi-directionally referenced by the service account) and not mounted by pods.
+
+In release 1.25 when this feature is Beta and enabled by default, delete secrets
+iff they have `used-since` label and the time duration since the date is over
+a sufficient period of time that can be configured by cluster admins.
```

Comment:
what do we intend the default to be here?

Reply:
4 months = 1 release? we neither have metric to know if this is a good value nor any constrain that is against this value.